Ecc siganture fallback for non-raw signature mechanisms

simo5 opened the pull request:
Add the ability to fallback to raw signature and separate digest calculation when tokens do not support Sign/Verify with digest calculation functions, but still support plain raw signatures. 

Fixes #168

Jakuje commented on tests/tecc:
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash -e
+# Copyright (C) 2022 Simo Sorce <simo@redhat.com>
```
nit: 2023 ;)

simo5 replied:
I didn't change contents, just move around stuff :)

Jakuje replied:
ok, missed that it was just moved around from the eccsha2. Thanks for clarification

Jakuje commented on tests/tecdh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -e
+# Copyright (C) 2022 Simo Sorce <simo@redhat.com>
```
here too 2023

simo5 replied:
Same, no new code, just moving stuff around

Jakuje commented on tests/tecc:
```diff
@@ -0,0 +1,57 @@
+                -digest sha256
+                -in ${RAND64FILE}
+                -rawin
+                -sigfile ${TMPPDIR}/sha256-ecdgstsig.bin'
```
would it make sense to give a shot to some other digests too, just to get more test coverage?

simo5 replied:
yes, I think that is a reasonable thing to do.

simo5 replied:
added sha384 and sha512, I do not think I can reasonably add more until we have a better way to exclude specific tests.
NSS softokn does not support SHA3 for example.

simo5 replied:
I guess if I make additional Digest fallback to OpenSSL digest I might be able to also to SHA3 ... Should we do that as part of this PR ?
It is possible that a token that does not support DigestSign APIs does not support Digest APIs at all after all ...

Jakuje replied:
Sounds like separate PR. SoftHSM will not support SHA3 probably either as far as I know. But its usage in the applications is not very wide so I think it should not be an issue.
In the next PR, I think this can be extended to allow fallback to OpenSSL implementation of digests or disabling this functionality through configuration file if this might be a concern to a certification.

Jakuje commented on tests/Makefile.am:
```diff
@@ -41,12 +41,13 @@ tmp.softhsm:
 	certs-softokn certs-softhsm-proxy \
-	eccsha2-softokn \
+	ecc-softokn ecc-softhsm-proxy \
+	ecdh-softokn \
```
the derive does not work with the softhsm (again?)

simo5 replied:
nope, that is why I split it off

Jakuje replied:
would be worth investigating if it is something in the softhsm, openssl or here to fix. Can you open a new issue for that?

Jakuje reviewed (commented):
looks much neater than I was expecting from what you were describing yesterday. Just nits. I pushed a change to try this branch on the libssh side to see if it will work:
https://gitlab.com/libssh/libssh-mirror/-/merge_requests/334
Edit: looks like it works as expected on the libssh side:
https://gitlab.com/jjelen/libssh-mirror/-/jobs/3687714156

simo5 commented on src/signature.c:
```diff
@@ -601,28 +606,60 @@ static int p11prov_sig_operate_init(P11PROV_SIG_CTX *sigctx, bool digest_op,
+                                  &session);
+        if (ret != CKR_OK) {
+            P11PROV_raise(sigctx->provctx, ret,
+                          "Failed to open fallback(%lx) session on slot %lu",
```
As silly as this may seem I actually went twice back and forth on this and actually decided for the no space version before pushing :-)

Jakuje commented on src/signature.c:
```diff
@@ -21,6 +21,8 @@ struct p11prov_sig_ctx {
 
     CK_RSA_PKCS_PSS_PARAMS pss_params;
+
+    bool digest_fallback;
```
last think after couple of reads of this PR, I am not sure about the naming of this variable. Its kind of fallback, but digest fallback sounds confusing/misleading, when this is basically flag saying "digest is calculated separately from the signature/verification operation" so my proposal would be "separate_digest". Or at least have on this line a comment describing what this variable/flag is doing.

simo5 replied:
I couldn't find a better name indeed, I used digest fallback, because the fallback is about separately calculating the digest, I could have called it mechanism_fallback, because it is about handling an unsupported mechanism.

simo5 replied:
See the fixup commit I added.

Jakuje replied:
After fixing the typo, looks good.

simo5 commented:
@Jakuje luckily I tried to do the digest fallback and after 2 attempts I realized that:

a) for digests fallbacks are not necessary at all because openssl providers list them individually and we already have code to expose them only if the token supports them (although we may still have an issue if we have multiple token and they differ in capabilities, I think this is for us to sort at a later time if it happens).

b) openssl already know how to fallback through providers if one does not support a digest.

I used these facts to add digest fallback capability by changing how the signature fallback works from trying to use directly p11prov_DigestInit and friends, to instead use EVP_Digest and asking to "prefer" the pkcs11 provider via "?provider=pkcs11".

I pushed this as an additional commit for now. But if you agree this should go all in this PR I will rebase and merge the initial implementation with the refactoring one in a single commit instead then split the sha3 test into the tests commit and add also sha3-384 and sha3-512.

Jakuje commented:
I am worried a bit with this direction that we will lose the ability to control which device will perform the digest operation in case there would be more pkcs11 devices. Going through the the OpenSSL might lead to using one device for the signature and completely other for the digest. If we are going to mix up things, I would rather see the OpenSSL internal implementation than some "random" device used and probably have some control over the behavior through some configuration.

simo5 commented:
> 

Uhmm... that may require us to try to do things internally and then call OpenSSL with provider=default ... it is possible but will double the code.

Unless we assume that if we are doing fallback we do not care for digesting on our own device at all and we always unconditionally defer to openssl default provider?

This would come back to byte us if we want to use FIPS devices that support hashes but not the combined operation on the token ... as we'd ask openssl default provider explicitly even though we can keep the operation on a FIPS validated device ...

beldmit commented:
I'd say that the fetching and its parameters belong to application logic and shouldn't be done on the provider logic.
Is it possible to add some more properties to algorithm objects, e.g. provider=pkcs11,driver=mycooltoken?

dengert commented:
https://www.rfc-editor.org/rfc/rfc7512 defines the PKCS11 URI that includes "module-name" and "module-path"  so provider could scan for this to get the PKCS11 module to load.

simo5 commented:
> https://www.rfc-editor.org/rfc/rfc7512 defines the PKCS11 URI that includes "module-name" and "module-path" so provider could scan for this to get the PKCS11 module to load.

There is no URI for selecting ciphers, only to reference objects ...

simo5 commented:
Ok it seem that if we can use properties to prefer "same" slot (passing ?pkcs11.slot=number) we can use the fallback.

There is still a bad case where we have multiple slots in a token and they have different capabilities. I think that means we may have to build fallbacks for pretty much all mechanisms ... but I think I want to handle that later.

simo5 commented:
@Jakuje this latest version is now able to prefer pkcs11 tokens but fall back to openssl if the digest is not ssupported at all.

If the pkcs11 provider is chosen we tell it what slotid to use (the same on which the key which we'll later us for signing is stored).

I think this should clear your reservations ?

beldmit commented on src/signature.c:
```diff
@@ -92,12 +100,27 @@ static void *p11prov_sig_dupctx(void *ctx)
 
     /* This is not really funny. OpenSSL by default assumes contexts with
-     * operations in flight can be easily duplicated, with all the
+      operations in flight can be easily duplicated, with all the
```
Nit: `*` removed by accident

Jakuje commented on src/signature.c:
```diff
@@ -739,6 +847,64 @@ static int p11prov_sig_digest_update(P11PROV_SIG_CTX *sigctx,
+
+done:
+    p11prov_sig_freectx(subctx);
+    return CKR_OK;
```
this should be `return ret` (with setting the `ret = CKR_OK` before the label) otherwise we are returning CKR_OK even if either of the last two actions fail.

simo5 replied:
good catch, changing to `return ret;`, and having *all* error conditions `goto done;`

simo5 commented:
I fixed the issues you both reported, I also added a little bit of refactoring with the return errors because I found an issue where I mixed ossl error and pkcs11 errors and that caused a failure. It took so much time to find out what I did that I decided to decrease the level of mixing removing it from internal functions I touched and pushing it to the boundary (the functions called directly by openssl).

I also rebases on top of latest main.

simo5 commented on src/signature.c:
```diff
@@ -92,8 +100,23 @@ static void *p11prov_sig_dupctx(void *ctx)
+    if (sigctx->mechanism_fallback) {
+        int err;
+        newctx->mechanism_fallback = EVP_MD_CTX_new();
+        if (!newctx->mechanism_fallback) {
```
I prefer the if (!var) format, but do not enforce it.

dengert commented:
Just a comment... @simo5 @Jakuje  you keep referring to the tokens support things. There is the PKCS11 module that plus the token. No smartcards look anything list PKCS11 In the OpenSC case the PKCS11 module links in libopensc that has PKCS15 interface as well an card drivers that deal with ISO7816-4  APDUs to talk to the token.

simo5 commented:
@dengert as mentioned, the caching was all of the problem. The token had *no* keys, but because of caching it looked like it did.
Once I removed caching from the config it all became clear. I since then generated a new RSA key in 9c and all the operations work as expected.

Btw in term of talking about "pkcs11 token", I know that tokens have pkcs#11 as an interface only and underneath there can be anything else, but that is just an abstraction the provider doesn't care about. From the PoV of the provider, pkcs#11 is the interface and tokens are pkcs#11 tokens.
